internal/controller/replication.storage/volumegroupreplication_controller.go

```diff
@@ -18,36 +18,772 @@ package controller
+			vgrContentObj.Annotations[volumeGroupReplicationRef] = vgrRef
+		}
+
+		if vgrContentObj.Spec.VolumeGroupReplicationRef == nil {
+			vgrContentObj.Spec.VolumeGroupReplicationRef = &corev1.ObjectReference{
+				APIVersion: vgr.APIVersion,
+				Kind:       vgr.Kind,
+				Name:       vgr.Name,
+				Namespace:  vgr.Namespace,
+				UID:        vgr.UID,
+			}
+		}
+
+		vgrContentObj.Spec.Source = replicationv1alpha1.VolumeGroupReplicationContentSource{
+			VolumeHandles: pvHandlesList,
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+func (r *VolumeGroupReplicationReconciler) cleanupVGRContent(vgr *replicationv1alpha1.VolumeGroupReplication,
+	vgrContentObj *replicationv1alpha1.VolumeGroupReplicationContent) error {
+	req := types.NamespacedName{Name: vgr.Spec.VolumeGroupReplicationContentName}
+	err := r.Client.Get(r.ctx, req, vgrContentObj)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			r.log.Error(err, "failed to fetch dependent volumeGroupReplicationContent resource")
+			return err
+		}
+		r.log.Info("volumeGroupReplicationContent resource not found")
+	} else {
+		err = r.Client.Delete(r.ctx, vgrContentObj)
+		if err != nil {
+			r.log.Error(err, "failed to delete dependent volumeGroupReplicationContent resource")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *VolumeGroupReplicationReconciler) createOrUpdateVolumeReplicationCR(vgr *replicationv1alpha1.VolumeGroupReplication,
+	vrObj *replicationv1alpha1.VolumeReplication) error {
+	apiGroup := "replication.storage.openshift.io"
+	_, err := controllerutil.CreateOrUpdate(r.ctx, r.Client, vrObj, func() error {
+		if vrObj.CreationTimestamp.IsZero() {
+			vrObj.Annotations = map[string]string{
+				volumeGroupReplicationRef: fmt.Sprintf("%s/%s", vgr.Namespace, vgr.Name),
+			}
+			vrObj.Spec = replicationv1alpha1.VolumeReplicationSpec{
+				VolumeReplicationClass: vgr.Spec.VolumeReplicationClassName,
+				DataSource: corev1.TypedLocalObjectReference{
+					APIGroup: &apiGroup,
+					Kind:     vgr.Kind,
+					Name:     vgr.Name,
+				},
+			}
+		}
+
+		vrObj.Spec.AutoResync = vgr.Spec.AutoResync
+		vrObj.Spec.ReplicationState = vgr.Spec.ReplicationState
+
+		return controllerutil.SetOwnerReference(vgr, vrObj, r.Scheme)
+	})
+
+	return err
+}
+
+func (r *VolumeGroupReplicationReconciler) cleanupVR(vgr *replicationv1alpha1.VolumeGroupReplication,
+	vrObj *replicationv1alpha1.VolumeReplication) error {
+	req := types.NamespacedName{Name: vgr.Spec.VolumeReplicationName, Namespace: vgr.Namespace}
+	err := r.Client.Get(r.ctx, req, vrObj)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			r.log.Error(err, "failed to fetch dependent volumeReplication resource")
+			return err
+		}
+		r.log.Info("volumeReplication resource not found")
+	} else {
+		err = r.Client.Delete(r.ctx, vrObj)
+		if err != nil {
+			r.log.Error(err, "failed to delete dependent volumeReplication resource")
+			return err
+		}
+	}
+
+	return nil
```

Comment:
There's no need for else block here.
```suggestion
	} 
		err = r.Client.Delete(r.ctx, vrObj)
		if err != nil {
			r.log.Error(err, "failed to delete dependent volumeReplication resource")
			return err
		}

	return nil
```

Reply:
@Rakshith-R ^^
We can avoid the `else` statement by adding a `return nil` at the end of `if` block, if that seems better?

Reply:
> @Rakshith-R ^^ We can avoid the `else` block by adding a `return nil` at the end of `if` block, if that seems better?

yes, let's do it this way.

The current flow makes it a bit confusing.

Reply:
Done